src/io/flutter/run/common/CommonTestConfigUtils.java

```diff
@@ -26,45 +32,76 @@
     return StringUtil.trimTrailing(
       url.replaceFirst("http:", "ws:"), '/') + "/ws";
   }
 
+  public abstract TestType asTestCall(@NotNull PsiElement element);
+
+  /**
+   * Gets the elements from the outline that are runnable tests.
+   */
+  protected Map<DartCallExpression, TestType> getTestsFromOutline(@NotNull PsiFile file) {
+    final Project project = file.getProject();
+    final FlutterOutline outline = OpenEditorOutlineService.getInstance(project).get(file.getVirtualFile().getPath());
+    final Map<DartCallExpression, TestType> callToTestType = new HashMap<>();
+    if (outline != null) {
+      visit(outline, callToTestType, file);
+    }
+    return callToTestType;
+  }
+
+  private void visit(@NotNull FlutterOutline outline, @NotNull Map<DartCallExpression, TestType> callToTestType, @NotNull PsiFile file) {
+    @NotNull final PsiElement element = Objects.requireNonNull(file.findElementAt(outline.getOffset()));
+    if (outline.getDartElement() != null) {
+      switch (outline.getDartElement().getKind()) {
+        case UNIT_TEST_GROUP:
+          // We found a test group.
+          callToTestType.put(DartSyntax.findClosestEnclosingFunctionCall(element), TestType.GROUP);
+          break;
+        case UNIT_TEST_TEST:
+          // We found a unit test.
+          callToTestType.put(DartSyntax.findClosestEnclosingFunctionCall(element), TestType.SINGLE);
+          break;
+        default:
+          // We found no test.
+          break;
+      }
+    }
+    if (outline.getChildren() != null) {
+      for (FlutterOutline child : outline.getChildren()) {
+        visit(child, callToTestType, file);
+      }
+    }
+  }
+
   @VisibleForTesting
   public boolean isMainFunctionDeclarationWithTests(@NotNull PsiElement element) {
     if (DartSyntax.isMainFunctionDeclaration(element)) {
-      final PsiElementProcessor.FindElement<PsiElement> processor =
-        new PsiElementProcessor.FindElement<PsiElement>() {
-          @Override
-          public boolean execute(@NotNull PsiElement element) {
-            final TestType type = findNamedTestCall(element);
-            return type == null || setFound(element);
-          }
-        };
-
-      PsiTreeUtil.processElements(element, processor);
-      return processor.isFound();
+      final Map<DartCallExpression, TestType> callToTestType = getTestsFromOutline(element.getContainingFile());
+      return !callToTestType.isEmpty();
     }
-
     return false;
   }
 
   @Nullable
   protected TestType findNamedTestCall(@NotNull PsiElement element) {
     if (element instanceof DartCallExpression) {
       final DartCallExpression call = (DartCallExpression)element;
-      for (TestType type : TestType.values()) {
-        if (type.matchesFunction(call)) return type;
+      final Map<DartCallExpression, TestType> callToTestType = getTestsFromOutline(element.getContainingFile());
+
+      if (callToTestType.containsKey(call)) {
```

Comment:
the call to containsKey isn't needed. Just return the value regardless of if it is null.

Reply:
Done.